cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionFilter.java

```diff
@@ -0,0 +1,245 @@
+
+  // we only allow creating an filter through the builder.
+  private PartitionFilter(Map<String, Condition<? extends Comparable>> conditions) {
+    this.conditions = conditions;
+  }
+
+  /**
+   * This should be used for inspection or debugging only.
+   * To match this filter, use the {@link #match} method.
+   *
+   * @return the individual conditions of this filter.
+   */
+  public Map<String, Condition<? extends Comparable>> getConditions() {
+    return ImmutableMap.copyOf(conditions);
+  }
+
+  /**
+   * @return the condition for a particular field.
+   */
+  public Condition<? extends Comparable> getCondition(String fieldName) {
+    return conditions.get(fieldName);
+  }
+
+  /**
+   * Match this filter against a partition key. The key matches iff it matches all conditions.
+   *
+   * @throws java.lang.IllegalArgumentException if one of the field types in the partition key are incompatible
+   */
+  public boolean match(PartitionKey partitionKey) {
+    for (Map.Entry<String, Condition<? extends Comparable>> condition : getConditions().entrySet())  {
+      Comparable value = partitionKey.getField(condition.getKey());
+      if (value == null || !condition.getValue().match(value)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public String toString() {
+    return conditions.values().toString();
+  }
+
+  /**
+   * Use this to create PartitionFilters.
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * A builder for partition filters.
+   */
+  public static class Builder {
+    private final Map<String, Condition<? extends Comparable>> map = Maps.newLinkedHashMap();
+
+    /**
+     * Add a condition for a given field name with an inclusive lower and an exclusive upper bound.
+     * Either bound can be null, meaning unbounded in that direction. If both upper and lower bound are
+     * null, then this condition has no effect and is not added to the filter.
+     *
+     * @param field The name of the partition field
+     * @param lower the inclusive lower bound. If null, there is no lower bound.
+     * @param upper the exclusive upper bound. If null, there is no upper bound.
+     * @param <T> The type of the partition field
+     *
+     * @throws java.lang.IllegalArgumentException if the field name is null, empty, or already exists,
+     *         or if both bounds are equal (meaning the condition cannot be satisfied).
+     */
+    public <T extends Comparable<T>> Builder addRangeCondition(String field,
+                                                               @Nullable T lower,
+                                                               @Nullable T upper) {
+      Preconditions.checkArgument(field != null && !field.isEmpty(), "field name cannot be null or empty.");
+      Preconditions.checkArgument(lower == null || !lower.equals(upper), "Unsatisfiable condition: " +
```

Comment:
Actually after discussing in person, it seems more reasonable to allow these conditions. The upper and lower bound may be programmatic parameters in some code that generates filter. It is valid to pass a condition that can be satisfied in general, but can't under current context.